test/common/run-tests

```diff
@@ -301,139 +315,135 @@ def detect_tests(test_files, image, opts):
+                if test_machine is not None and not opts.machine:
+                    # unassign from global machine
+                    global_machines[test_machine].running_test = None
 
                     # sometimes our global machine gets messed up; also, tests that time out don't run cleanup handlers
                     # restart it to avoid an unbounded number of test retries and follow-up errors
                     if not opts.machine and (poll_result == 124 or (retry_reason and "test harness" in retry_reason)):
                         # try hard to keep the test output consistent
-                        sys.stderr.write("\nRestarting global machine %s\n" % test.machine_id)
+                        sys.stderr.write("\nRestarting global machine %s\n" % test_machine)
                         sys.stderr.flush()
-                        batch_tests[test.machine_id]["machine"].reset()
+                        global_machines[test_machine].reset()
 
                 # run again if needed
                 if retry_reason:
-                    if test.machine_id is not None:
-                        batch_tests[test.machine_id]["tests"].insert(0, test)
-                        serial_remaining = sum([len(batch_tests[x]["tests"]) for x in batch_tests])
+                    if test.nondestructive:
+                        serial_tests.insert(0, test)
                     else:
                         parallel_tests.insert(0, test)
 
-                if test.machine_id is not None:
-                    batch_tests[test.machine_id]["working"] = None
+        if opts.machine:
+            if not running_tests and serial_tests:
+                test = serial_tests.pop(0)
+                logging.debug(f"Static machine is free, assigning next test {test}")
+                test.assign_machine(-1, opts.machine, opts.browser)
+                test.start()
+                running_tests.append(test)
+                made_progress = True
+        else:
+            # find free global machines, and either assign a new serial test, or kill them to free resources
+            for (idx, machine) in enumerate(global_machines):
+                if machine.is_available():
+                    if serial_tests:
+                        test = serial_tests.pop(0)
+                        logging.debug(f"Global machine {idx} is free, assigning next test {test}")
+                        machine.running_test = test
+                        test.assign_machine(idx, machine.ssh_address, machine.web_address)
+                        test.start()
+                        running_tests.append(test)
+                    else:
+                        logging.debug(f"Global machine {idx} is free, and no more serial tests; killing")
+                        machine.kill()
+
+                    made_progress = True
+
+        # fill the remaining available job slots with parallel tests
+        while parallel_tests and len(running_tests) < opts.jobs:
```

Comment:
The logic here is funny.  For some reason I kinda thought that we'd rather look at the number of running serial machines before going to start parallel jobs, but since the code above pretty much ensures that there is the exact number of tests running as there are serial machines, then this is equivalent...

So fine, but not what I was expecting to see.  Probably better, though, since we don't need to keep track of the number of running "parallel" tests separately.